### PR TITLE
[HPRO-1216] Henry Ford Repair Cron.

### PR DESCRIPTION
**Pull request by Ehaic**
Added missing participants.

| Q                   | A
| ------------------- | --------------
| Target Release?     | Next available 
| Bug fix?            | ❌ 
| New feature?        | ❌ 
| Database migration? | ❌ 
| New configuration?  | ❌ 
| Composer updates?   | ❌ 
| NPM updates?        | ❌ 
| Jira Ticket(s)      | HPRO-1216 

### Summary

Adds a cron to repair participants to new sites.

### Screenshots 

[HPRO-1216]: https://precisionmedicineinitiative.atlassian.net/browse/HPRO-1216?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Service/HFHRepairService.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                $CSVArray = array_slice($CSVArray, 1 + $count);
+                $CSVArray = array_merge([implode(',', $headers)], $CSVArray);
+                file_put_contents('src/Cache/HFSitePairing.csv', implode("\r\n", $CSVArray));
+                break;
+            }
+        }
+        $this->logger->info('100 records processed');
+    }
+
+    private function repairParticipantSite($participantId, $currentSite, $repairSite)
```
Comment (shyamzcode) [inline src/Service/HFHRepairService.php:52]: You should remove the site prefix "hpo-site-" while comparing and while inserting into the db.

**Review by shyamzcode — CHANGES_REQUESTED** [review]
I see the repair for only orders table we also need to do it for the evaluations table as well.

File: src/Service/HFHRepairService.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        $evaluation = $this->em->getRepository(Measurement::class)->findBy(['participantId' => $participantId, 'finalizedSite' => $currentSite]);
+        $order = $this->em->getRepository(Order::class)->findBy(['participantId' => $participantId, 'finalizedSite' => $currentSite]);
+        if (count($evaluation) == 0) {
+            throw new \Exception("No order found for participant $participantId at site $currentSite");
+        }
+        if (count($order) == 0) {
+            throw new \Exception("No measurements found for participant $participantId at site $currentSite");
+        }
+        foreach ($order as $order) {
+            $order->setFinalizedSite($repairSite);
```
Comment (shyamzcode) [inline src/Service/HFHRepairService.php:68]: We should always log the ids when something is being created or updated in the db . Could you log the order id here? You could do that using. 
`$this->loggerService->log(Log::ORDER_EDIT, $order->getId());`

File: src/Service/HFHRepairService.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        $currentSite = str_replace('hpo-site-', '', $currentSite);
+        $evaluation = $this->em->getRepository(Measurement::class)->findBy(['participantId' => $participantId, 'finalizedSite' => $currentSite]);
+        $order = $this->em->getRepository(Order::class)->findBy(['participantId' => $participantId, 'finalizedSite' => $currentSite]);
+        if (count($evaluation) == 0) {
+            throw new \Exception("No order found for participant $participantId at site $currentSite");
+        }
+        if (count($order) == 0) {
+            throw new \Exception("No measurements found for participant $participantId at site $currentSite");
+        }
+        foreach ($order as $order) {
```
Comment (shyamzcode) [inline src/Service/HFHRepairService.php:67]: Looks like both variables are same. Could you use $orders and iterate. Like
`foreach ($orders as $order)`
You need to first change the variable name $order to $orders in line 60.

File: src/Service/HFHRepairService.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            $order->setFinalizedSite($repairSite);
+        }
+        foreach ($evaluation as $measurement) {
+            $measurement->setFinalizedSite($repairSite);
```
Comment (shyamzcode) [inline src/Service/HFHRepairService.php:71]: Same as above could you log the measurement id?

**Review by shyamzcode — CHANGES_REQUESTED** [review]
Overall looks good! Also, you need to register the cron job in the cron.yaml.dist file in order for this cron job to run. Could you add this job in the cron.yaml.dist file and set the time to run every 5 mins?

File: src/Service/HFHRepairService.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                fclose($fhandle);
+                $CSVArray = file_get_contents('src/Cache/HFSitePairing.csv');
+                $CSVArray = explode("\r\n", $CSVArray);
+                $CSVArray = array_slice($CSVArray, 1 + $count);
+                $CSVArray = array_merge([implode(',', $headers)], $CSVArray);
+                file_put_contents('src/Cache/HFSitePairing.csv', implode("\r\n", $CSVArray));
+                break;
+            }
+        }
+        $this->logger->info('100 records processed');
```
Comment (thornview) [inline src/Service/HFHRepairService.php:50]:
```suggestion
        $this->logger->info($repairLimit . ' records processed');
```

File: src/Service/HFHRepairService.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    private function repairParticipantSite($participantId, $currentSite, $repairSite)
```
Comment (thornview) [inline src/Service/HFHRepairService.php:53]: Use type hinting for the parameters and return value.

File: src/Service/HFHRepairService.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    private EntityManagerInterface $em;
+    private LoggerInterface $logger;
+
+    public function __construct(EntityManagerInterface $em, LoggerInterface $logger)
+    {
+        $this->em = $em;
+        $this->logger = $logger;
+    }
+
+    public function repairHFHParticipants($repairLimit = 100): void
```
Comment (thornview) [inline src/Service/HFHRepairService.php:22]: Add type hint for parameter.

File: src/Service/HFHRepairService.php (hunk 3 above)
Comment (thornview) [reply src/Service/HFHRepairService.php:67]: I agree with Shyam here.  Do not reuse variable names.  It creates some confusing code.

**Review by thornview — CHANGES_REQUESTED** [review]

**Pull request closed without merging**